app/src/main/java/at/tugraz/ist/swe/note/NoteStorage.java

```diff
@@ -0,0 +1,93 @@
+
+    public NoteStorage(DatabaseHelper databaseHelper) {
+        this.databaseHelper = databaseHelper;
+    }
+
+    static private String convertDateToString(Date date) {
+        return DATE_FORMAT.format(date);
+    }
+
+    static private Date convertStringToDate(String date) {
+        try {
+            return DATE_FORMAT.parse(date);
+        } catch(java.text.ParseException e) {
+            return null;
+        }
+    }
+
+    /**
+     * @return Returns true if note was inserted, false otherwise.
+     */
+    public boolean save(Note note) {
+        if(note.exists()) {
+            update(note);
+            return false;
+        } else {
+            insert(note);
+            return true;
+
+        }
+    }
+
+    public Note findById(long id) throws NotFoundException {
+        SQLiteDatabase database = databaseHelper.getReadableDatabase();
+        String selection = DatabaseHelper.NOTE_COLUMN_ID + " = ?";
+        String[] selectionArgs = {String.valueOf(id)};
+        Cursor cursor = database.query(DatabaseHelper.NOTE_TABLE_NAME, null, selection, selectionArgs, null, null, null);
+        if(cursor.getCount() != 1) {
+            throw new NotFoundException(id);
+        }
+        Note note = new Note();
+        cursor.moveToFirst();
+        note.setId(cursor.getLong(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_ID)));
+        note.setTitle(cursor.getString(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_TITLE)));
+        note.setContent(cursor.getString(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_CONTENT)));
+        note.setCreatedDate(convertStringToDate(cursor.getString(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_CREATED_DATE))));
+        note.setChangedDate(convertStringToDate(cursor.getString(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_CHANGED_DATE))));
+        note.setPinned(cursor.getInt(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_PINNED)));
+        note.setRemoved(cursor.getInt(cursor.getColumnIndex(DatabaseHelper.NOTE_COLUMN_REMOVED)) > 0);
+        cursor.close();
+        return note;
+    }
+
+    private ContentValues getContentValues(Note note) {
+        ContentValues values = new ContentValues();
+        values.put(DatabaseHelper.NOTE_COLUMN_TITLE, note.getTitle());
+        values.put(DatabaseHelper.NOTE_COLUMN_CONTENT, note.getContent());
+        values.put(DatabaseHelper.NOTE_COLUMN_PINNED, note.getPinned());
+        values.put(DatabaseHelper.NOTE_COLUMN_CREATED_DATE, convertDateToString(note.getCreatedDate()));
+        values.put(DatabaseHelper.NOTE_COLUMN_CHANGED_DATE, convertDateToString(note.getChangedDate()));
+        values.put(DatabaseHelper.NOTE_COLUMN_REMOVED, note.isRemoved());
+        return values;
+    }
+
+    private void insert(Note note) {
+        SQLiteDatabase database = databaseHelper.getWritableDatabase();
+        long id = database.insert(DatabaseHelper.NOTE_TABLE_NAME, null, getContentValues(note));
```

Comment:
update changes the 'changeDate'. insert doesn't set the date. it should be always the same logic.

Reply:
Insert does not need to set it because the constructor of Note sets it.

Reply:
but why do we set it once in the constructor and once in the store? which class is responsible for the date value? node or noteStorage?

Reply:
I moved the computation of the createdDate and changedDate to the database (defined as default values of the columns) and it is now only computed, when the database is really inserting or updating the note.

Reply:
perfect